extensions/amp-video/0.1/test/test-mixins.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or videoImplied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpVideo} from '../amp-video';
+import {Services} from '../../../../src/services';
+import {VideoElementMixin, MediaPoolVideoMixin} from '../mixins';
+
+
+describes.fakeWin('amp-video mixins', {}, env => {
+  let win;
+  let doc;
+  let element;
+  let videoImpl;
+
+  const vsync = {
+    mutate(fn) {
+      fn();
+    },
+  };
+
+  beforeEach(() => {
+    win = env.win;
+    doc = win.document;
+
+    element = doc.createElement('amp-video');
+    videoImpl = {
+      element,
+      getVsync() {
+        return vsync;
+      },
+    };
+  });
+
+  describe('VideoElementMixin', () => {
+
+    it('should create and append a video as base node', () => {
+      const mixin = new VideoElementMixin(videoImpl);
+      const baseNode = mixin.createBaseNode();
+
+      expect(baseNode.tagName).to.equal('VIDEO');
+      expect(element.firstElementChild).to.equal(baseNode);
+    });
+  });
+
+  describe('MediaPoolVideoMixin', () => {
+    let pool;
+    let hlsStub;
+
+    beforeEach(() => {
+      pool = {
+        play: sandbox.spy(),
+        pause: sandbox.spy(),
+      };
+
+      sandbox.stub(Services, 'mediaPoolFor')
+          .returns(Promise.resolve(pool));
+    });
+
+    it('should create and append a video as base node', () => {
```

Comment:
Should it?  Can we avoid creating an actual video element and *only* show the placeholder in its place, until a video is allocated from the pool?  I'd love to avoid this case:

 - A document that has 100 videos
- It's managed by `MediaPool`, only the 8 created by `MediaPool` will ever be used
- **We still create 100 video elements (worse if we actually set the src) to be used only as placeholders**

Reply:
It shouldn't, but now it does. See [comment above.](https://github.com/ampproject/amphtml/pull/14316#discussion_r178322919)

I see three steps into the process of fixing this:

1. Landing this PR
2. Change registration API so that the prototype node is passed to the media pool by `MediaPoolVideoMixin` and never gets appended in the document. This is already a big gain for performance, since the browser does not schedule any painting, etc.
3. Provide a synthetic object that satisfies the `Node` interface as a prototype node. This way we don't create actual elements.